gen3-client/g3cmd/upload-manifest.go

```diff
@@ -0,0 +1,209 @@
+					fmt.Printf("Error\n")
+					errors[i] = nil
+					completed++
+				}
+			}
+		}
+	}
+
+	t.Stop()
+	fmt.Printf("%d files uploaded.\n", len(reqs))
+}
+
+func init() {
+	var manifestPath string
+	var uploadPath string
+	var fileType string
+	var batch bool
+	var numParallel int
+
+	var uploadManifestCmd = &cobra.Command{
+		Use:   "upload-manifest",
+		Short: "upload files from a specified manifest",
+		Long: `Gets a presigned URL for a file from a GUID and then uploads the specified file.
+	Examples: ./gen3-client upload-manifest --profile user1 --manifest manifest.tsv --upload-path=files/ 
+	`,
+		Run: func(cmd *cobra.Command, args []string) {
+
+			request := new(jwt.Request)
+			configure := new(jwt.Configure)
+			function := new(jwt.Functions)
+
+			function.Config = configure
+			function.Request = request
+
+			var objects []ManifestObject
+
+			manifestFile, err := os.Open(manifestPath)
+			if err != nil {
+				panic(err)
+			}
+			defer manifestFile.Close()
+
+			switch {
+			case strings.EqualFold(filepath.Ext(manifestPath), ".tsv"):
+				r := csv.NewReader(manifestFile)
+				r.Comma = '\t'
+				for {
+					record, err := r.Read()
+					if err == io.EOF {
+						break
+					} else if err != nil {
+						log.Fatalf("TSV parse error\n")
+					}
+					objects = append(objects, ManifestObject{ObjectID: record[0], SubjectID: record[1]})
+				}
+			case strings.EqualFold(filepath.Ext(manifestPath), ".json"):
+				manifestBytes, err := ioutil.ReadFile(manifestPath)
+				if err != nil {
+					log.Fatalf("Failed reading manifest %s, %v\n", manifestPath, err)
+				}
+				json.Unmarshal(manifestBytes, &objects)
+			default:
+				log.Fatalf("Unsupported manifast format")
+				return
+			}
+
+			if batch {
+				reqs := make([]*http.Request, 0)
+				for _, object := range objects {
+					endPointPostfix := "/user/data/upload/" + object.ObjectID
+					respURL, err := function.DoRequestWithSignedHeader(profile, "", endPointPostfix)
+
+					data, err := ioutil.ReadFile(uploadPath + "/" + object.SubjectID)
+					if err != nil {
+						fmt.Println(err.Error())
+						break
+					}
+					body := bytes.NewBufferString(string(data[:]))
+					contentType := "application/json"
+					if fileType == "tsv" {
+						contentType = "text/tab-separated-values"
+					}
+					req, _ := http.NewRequest(http.MethodPut, respURL, body)
+					req.Header.Set("content_type", contentType)
+					reqs = append(reqs, req)
+				}
+				batchUpload(numParallel, reqs)
+			} else {
+				for _, object := range objects {
+					endPointPostfix := "/user/data/upload/" + object.ObjectID
+					respURL, err := function.DoRequestWithSignedHeader(profile, "", endPointPostfix)
+
+					if err != nil {
+						if strings.Contains(err.Error(), "The provided guid") {
+							log.Printf("Upload error: %s\n", err)
+						} else {
+							log.Fatalf("Fatal upload error: %s\n", err)
+						}
+					} else {
+						uploadFile(object.ObjectID, uploadPath+"/"+object.SubjectID, fileType, respURL)
```

Comment:
Lets use object.ObjectID for the path instead of SubjectID? This way a user can download a manifest, and then reupload again without having to rename files...

Reply:
I'll change this part because it is actually confusing. I was not suppose to use SubjectID here. Cause what I really need is the file name for upload.

In `download-manifest` the ObjectID is the UUID and SubjectID is never used. In my `upload-manifest` here the ObjectID is suppose to be the UUID again and the SubjectID is the filename to be uploaded to that UUID. I'll change the name for `UploadManifestObject` to make it more meaningful. I'm following the example manifest file format as shown in https://gen3.org/resources/user/gen3-client/

Reply:
Ah, those aren't official manifests... the only ones we have so far are from the data explorer that you can download through Windmill. Lets only write code to support those, or identify enhancements that are needed in them.